pkg/controllers/user/helm/state.go

```diff
@@ -0,0 +1,133 @@
+package helm
+
+import (
+	"context"
+	"fmt"
+	"github.com/rancher/norman/types/convert"
+	mcapp "github.com/rancher/rancher/pkg/controllers/user/multiclusterapp"
+	util "github.com/rancher/rancher/pkg/controllers/user/workload"
+	"github.com/rancher/rancher/pkg/namespace"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	pv3 "github.com/rancher/types/apis/project.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"strings"
+)
+
+const MultiClusterAppIDSelector = "mcapp"
+
+func StartStateCalculator(ctx context.Context, cluster *config.UserContext) {
+	apps := cluster.Management.Project.Apps("")
+	w := &AppStateCalculator{
+		AppLister:        apps.Controller().Lister(),
+		Apps:             apps,
+		MultiClusterApps: cluster.Management.Management.MultiClusterApps(""),
+	}
+	w.workloadLister = util.NewWorkloadController(ctx, cluster.UserOnlyContext(), w.sync)
+	apps.Controller().AddHandler(ctx, "app-state-controller", w.syncAppState)
+}
+
+type AppStateCalculator struct {
+	Apps             pv3.AppInterface
+	AppLister        pv3.AppLister
+	MultiClusterApps v3.MultiClusterAppInterface
+	workloadLister   util.CommonController
+}
+
+func (s *AppStateCalculator) sync(key string, obj *util.Workload) error {
+	if obj == nil {
+		// possible for workload to be already deleted?
+		return nil
+	}
+	if label, ok := obj.Labels[appLabel]; ok {
+		apps, err := s.AppLister.List("", labels.NewSelector())
```

Comment:
looks like here we are enqueueing all the apps? We should just get the app by id

Reply:
To get the app namespace, get the workload namespace, get the project it belongs to. The project name correlates to app namespace on rancher